### PR TITLE
NCMEC Reporting

### PR DESCRIPTION
**Pull request by peterjan**
# PULL REQUEST

## Overview

This PR adds an extra module that reports CSAM abuse emails to NCMEC.

NOTE: this PR is blocked until we can provide IP and date of upload(er)
## Example for Visual Changes
N/A

## Checklist
Review and complete the checklist to ensure that the PR is complete before assigned to an approver.
 - [x] All new methods or updated methods have clear docstrings
 - [x] Testing added or updated for new methods
 - [ ] Verify if any changes impact the WebPortal Health Checks
 - [ ] Approriate documentation updated
 - [ ] Changelog file created

### CODE REVIEW

File: main.go
Comment (ChrisSchinnerl) [inline main.go:120]: I would go for an approach like we do with TUS and Jaeger where you can enable or disable ncmec reporting with an env var but if it is enabled, we won't boot if we encounter an error due to missing env vars.

File: email/reporter.go
Comment (ChrisSchinnerl) [inline email/reporter.go:115]: Could do this before the `if` to avoid calling it in 2 places.

File: email/reporter.go
Comment (ChrisSchinnerl) [inline email/reporter.go:144]: What happens when the report isn't closed? Or you try to open the same report twice?

File: email/reporter.go
Comment (ChrisSchinnerl) [inline email/reporter.go:215]: I assume this is supposed to be set to the time we received the email?

File: email/reporter_test.go
Comment (ChrisSchinnerl) [inline email/reporter_test.go:45]: commented out code

File: email/reporter.go
Comment (ChrisSchinnerl) [inline email/reporter.go:226]: Is there no way to report as a company?

If not then `SkynetLabs` and `Inc.` might be better?

File: email/reporter.go
Comment (ChrisSchinnerl) [inline email/reporter.go:227]: Are we still using `siasky.net` for abuse?

File: email/ncmec.go
Comment (ro-tex) [inline email/ncmec.go:34]: This one could use a comment explaining what it does.

File: email/ncmec.go
Comment (ro-tex) [inline email/ncmec.go:235]: nit: redundant

File: email/ncmec.go
Comment (ro-tex) [inline email/ncmec.go:272]: nit: redundant

File: email/ncmec_test.go
Comment (ro-tex) [inline email/ncmec_test.go:19]: forgotten debug

File: email/ncmec_test.go
Comment (ro-tex) [inline email/ncmec_test.go:95]: The seem to be using custom response codes. It would be nice to export those to constants with descriptive names. Otherwise, anyone trying to maintain this code will need to look up their help pages and go through all the research you did.

File: email/ncmec_test.go
Comment (ro-tex) [inline email/ncmec_test.go:106]: We have a mix of UTC and non-UTC times in the code. We should stick to UTC.

Screenshot 2022-03-25 at 11 18 14

File: email/reporter.go
Comment (ro-tex) [reply email/reporter.go:227]: We need all of these configurable. Independent portal operators should be able to plug in their credentials/identity and use this without forking the codebase.

File: email/reporter.go
Comment (ro-tex) [inline email/reporter.go:206]: The portal URL needs to be configurable.

Comment (ChrisSchinnerl) [reply email/reporter.go:206]: Good point. We should already have an env var for that anyway.

File: email/ncmec_test.go
Comment (ChrisSchinnerl) [reply email/ncmec_test.go:106]: Oh wow that's a lot of calls. Maybe we can add a helper method within the relevant package.

Comment (ro-tex) [reply email/ncmec_test.go:106]: I don't think that would be an improvement, tbh. When coding I would still type `time.Now().UTC()`, it's in my muscle memory. Then you'll ask me to use the helper in the code review... Just more time wasted?
Your call but I don't see that as an improvement.

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:115]: There are two different log levels, if it's fine by you I'd like to keep it the way it is.

If the abuse scanner is run in `debug` mode the modules log a message indicating they didn't found any "work".
If the abuse scanner is run in `info` mode it will only log a message when it's about to process work.

File: main.go
Comment (peterjan) [reply main.go:120]: yeah this didn't feel right, done 👍

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:144]: Well, nothing really happens, we create noise on their end. I was thinking about this when writing it as well, so I updated the flow a bit to improve it. 

I made it so we:
- try and finish an open report if we have its id
- try and finish a report if we have its id, even if something went wrong after opening it
- update the email with potential reporting errors on both the open and finish stage of the report

No need to overthink this any more than that for the time being I think. But up for review obv.

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:215]: I didn't care because the time we received the email is also hardly the incident date or time. In happy cases, the time we received the email and the time we pick it up differ at most 30s. I can try and see how hard it is to parse out the date from the email if you want.

File: email/reporter_test.go
Comment (peterjan) [reply email/reporter_test.go:45]: I want to leave that in as a placeholder.

File: database/abuseemail.go
Comment (peterjan) [reply database/abuseemail.go:161]: done

File: database/abusescanner.go
Comment (peterjan) [reply database/abusescanner.go:43]: I thought about that when writing it but you get an annoying cyclic reference. Because you still want to return the `DB` struct from the database package, and you want to use your test db from your database package. So that yields a cyclic ref. which is annoying.

I don't see the big deal with `database.NewTestDatabase()` anyway, I think it's nice and it works.

File: database/abusescanner.go
Comment (peterjan) [reply database/abusescanner.go:233]: I did it for readability because the comment above says
```
// unfinalized emails are either:
// - emails without csam that have not been finalized
// - emails with csm that have been reported but not yet finalized
```
so it matched 1-1, but updated it 👍

File: email/ncmec.go
Comment (peterjan) [reply email/ncmec.go:34]: updated

File: email/ncmec_test.go
Comment (peterjan) [reply email/ncmec_test.go:19]: I kind of want to leave these in... It's like a placeholder for when you want to run the tests against the actual NCMEC (test) API

File: email/ncmec_test.go
Comment (peterjan) [reply email/ncmec_test.go:106]: I'll update it to UTC everywhere, usually I just don't care in unit tests, it's an extra couple of characters to type and it's longer.

File: email/ncmec_test.go
Comment (peterjan) [reply email/ncmec_test.go:19]: closing this one in favour of https://github.com/SkynetLabs/abuse-scanner/pull/12#discussion_r834241200

File: database/abusescanner.go
Comment (ro-tex) [reply database/abusescanner.go:43]: Aw, that's a bummer. Sure, let's keep it here then. Not ideal but cycling deps are a PITA.

File: email/ncmec_test.go
Comment (ro-tex) [reply email/ncmec_test.go:106]: Yes, it's that important in unit tests but it does help.
I was more worried about `finalizer.go:127`.

File: email/ncmec_test.go
Comment (peterjan) [reply email/ncmec_test.go:95]: done 👍

File: email/ncmec_test.go
Comment (peterjan) [reply email/ncmec_test.go:106]: I'll resolve this, I agree with Ivo on the helper. This looks like a ton of occurrences but it's 70% testing and I think it's better to explicitly use `.UTC()`

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:206]: Made all of these configurable, was going to F/U with that but did it here.

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:227]: we are using `[report|abuse]@[portal]` for abuse, making these configurable indeed, was just going to F/U with that

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:226]: I'll make it `SkynetLabs Inc.`, no there is no way to report as a company, strangely enough..

File: email/reporter.go
Comment (ro-tex) [inline email/reporter.go:450]: nit: Shadowing types with local variable names is not great. It doesn't cause any direct problems here or in all the other cases throughout the PR but I wanted to mention it.

File: database/abusedb.go
Comment (peterjan) [reply database/abusedb.go:275]: yeah in mongo the $in is redundant if the type is an array

File: accounts/client.go
Comment (peterjan) [reply accounts/client.go:53]: I'll add it to the `README` this is consistent with the `BLOCKER_HOST` and `BLOCKER_PORT` env variables.

File: database/abusedb_test.go
Comment (peterjan) [reply database/abusedb_test.go:68]: lol never considered this, this thing is in my snippets so I copy-paste it all over, thanks!

File: database/abusedb_test.go
Comment (peterjan) [reply database/abusedb_test.go:211]: updated

File: database/abuseemail.go
Comment (peterjan) [reply database/abuseemail.go:157]: sorry, by fixing a merge conflict here I mistakenly got rid of this update I assume

File: database/reportdb.go
Comment (peterjan) [reply database/reportdb.go:35]: leftover, thanks

File: database/reportdb.go
Comment (peterjan) [reply database/reportdb.go:69]: Yeah, this is the thing we go back and forth on right 😛  I always suggest you do it this way in your PRs. I like this personally, if you get a pointer you should generally always check if it's nil or not and I like the fact that you, as a caller, don't have to think about database-specific errors like the mongo err no documents.

I know we could return a custom `ErrNoDocument` but at this point it would be inconsistent to handle it this way seeing as in all of the abuse repos I use the same pattern. I don't mind updating it if I stand alone on this but I'll do it in a F/U to tackle all occurrences.

But give it a chance, I think it's pretty clean and it reads nice and it's one less error to handle and you get around the awkward pattern of checking an error, seeing it's not nil and going "oh no", and then "oof" it's just an `ErrNoDocuments`... I mean you're getting a pointer, to me this is the equivalent of a `Promise<object|null>` in javascript for instance.

Comment (ro-tex) [reply database/reportdb.go:57]: Oh, haha, it's true that we've been over this. It would be nice to get some input from other team members here. cc @ChrisSchinnerl @MSevey 

I'm not convinced this approach is less work because I check for errors anyway and now I'll need to check for `nil` as well. I don't like that but I'm not some kind of hill to die on, so if the team likes the `nil` approach I can start using it. I agree that passing DB errors leads to leaky abstractions and that custom `not found` errors are a bit clumsy because you need to define them (I have a bunch of them in `accounts`).

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:450]: I updated it in one place where I could easily update it to `r`. It's a little bit annoying the module is called `Reporter` so I can't go with `r` here. I prefer `report report` over some weird looking variable name in this case. Changed it in the other place though.

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go:487]: Hm yeah so bit of a weird case here because I don't consider the database update a failure to open report, we go on and try to finish the report after this using the report id, if that succeeds everything's fine.

Comment (peterjan) [reply email/reporter.go:487]: I removed the return and added a comment to clarify

File: email/reporter.go
Comment (ro-tex) [inline email/reporter.go:479]: This update is different from the original behaviour. The previous version of the code logged this message regardless of the success of `UpdateReportNoLock` and logged the composed error. Are you sure we don't want to do that anymore?

File: email/reporter.go
Comment (ro-tex) [reply email/reporter.go]: This is great now. The comment explains why we're not returning the error and the `return` that's now gone no longer confuses the reader. Perfect.

**Review by ro-tex — APPROVED** [review]
LGTM.

We should continue the discussion about `nil` on `not found` on #core-dev after this is merged.

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go]: thanks, oversight

**Review by ro-tex — APPROVED** [review]

**Review by ro-tex — DISMISSED** [review]

File: email/reporter.go
Comment (peterjan) [reply email/reporter.go]: planning to F/U with this, lost track of this comment

**Review by ro-tex — APPROVED** [review]

File: database/reportdb.go
Comment (ChrisSchinnerl) [reply database/reportdb.go:57]: I'd stick to the conventions. e.g. `os` exposes `os.IsNotExist`. In this case I'd expose a `database.IsNotFound` that you can use.

This is clearer than knowing you have to check for `nil` even if there is no error. It's also not more lines of code. When you know there must be a document it's actually less verbose.

**Review by ChrisSchinnerl — APPROVED** [review]

**Pull request merged**
